File: day_07/day_07.go
package day_07

import (
	a "aoc_2024/aocutils"
	"strings"
)

func canCompute(target int, operands []int, val int, isPart1 bool) bool {
	if len(operands) == 0 {
		return val == target
	} 
	return canCompute(target, operands[1:], val + operands[0]					  , isPart1) ||
		   canCompute(target, operands[1:], val * operands[0]					  , isPart1) ||
		   canCompute(target, operands[1:], a.ToI(a.ToS(val) + a.ToS(operands[0])), isPart1) && !isPart1
}

func Day_07() {
	eqns := a.Reader(7)
	total1, total2 := 0, 0
	for _, e := range eqns {
		e := strings.Split(e, ": ")
		t, os := a.ToI(e[0]), a.MapToInt(strings.Split(e[1], " "))
		if canCompute(t, os, 0, true) {
			total1 += t
		}
		if canCompute(t, os, 0, false) {
			total2 += t
		}
	}

	a.PrintPartsInt(total1, total2)
}

File: day_07/day_07.py
import aocutils
from multiprocessing import Pool

def extract(x):
    val, operands = x.split(": ")
    return (int(val), list(map(int, operands.split(" "))), 0)

def canCompute(target, operands, val, part1):
    if not operands:
        return target == val
    
    isComputable = canCompute(target, operands[1:], val + operands[0]               , part1) \
                or canCompute(target, operands[1:], val * operands[0]               , part1) \
                or canCompute(target, operands[1:], int(str(val) + str(operands[0])), part1) and not part1
    
    return target if isComputable else 0

@aocutils.timeFunction
def main():
    pool = Pool()
    total1 = 0
    total2 = 0
    eqns = list(map(lambda x: extract(x), aocutils.readFile(7, isTest=False, toList=False)))

    total1 = sum(pool.starmap(canCompute, list(map(lambda x: x + (True,), eqns))))
    total2 = sum(pool.starmap(canCompute, list(map(lambda x: x + (False,), eqns))))

    aocutils.printParts(total1, total2)

if __name__ == "__main__":
    main()